src/opt/ReplaceHalfMaskedMemOps.cpp

```diff
@@ -0,0 +1,216 @@
+/*
+  Copyright (c) 2024, Intel Corporation
+
+  SPDX-License-Identifier: BSD-3-Clause
+*/
+
+#include "ReplaceHalfMaskedMemOps.h"
+
+#include <llvm/IR/IRBuilder.h>
+
+namespace ispc {
+
+bool lIsSecondHalfAllFalse(llvm::Value *mask) {
+    if (auto *CV = llvm::dyn_cast<llvm::ConstantVector>(mask)) {
+        auto N = CV->getType()->getNumElements();
+        for (auto i = N / 2; i < N; i++) {
+            llvm::Constant *E = CV->getAggregateElement(i);
+            if (!E || !llvm::isa<llvm::ConstantInt>(E) || !llvm::cast<llvm::ConstantInt>(E)->isZero()) {
+                // found non-zero element
+                return false;
+            }
+        }
+        for (auto i = 0; i < N / 2; i++) {
+            llvm::Constant *E = CV->getAggregateElement(i);
+            if (!E || !llvm::isa<llvm::ConstantInt>(E) || !llvm::cast<llvm::ConstantInt>(E)->isOne()) {
+                // found non-one element
+                return false;
+            }
+        }
+        return true;
+    }
+    return false;
+}
+
+llvm::Value *lMergeVectors(llvm::IRBuilder<> &B, llvm::Value *firstVector, llvm::Value *secondVector,
+                           llvm::Twine &name) {
+    auto *firstVecType = llvm::cast<llvm::VectorType>(firstVector->getType());
+    auto *secondVecType = llvm::cast<llvm::VectorType>(secondVector->getType());
+    unsigned N = firstVecType->getElementCount().getKnownMinValue();
+
+    assert(firstVecType == secondVecType);
+
+    // Create a vector of indices to merge two vectors
+    std::vector<llvm::Constant *> indices;
+    for (unsigned i = 0; i < N * 2; ++i) {
+        indices.push_back(llvm::ConstantInt::get(B.getInt32Ty(), i));
+    }
+    llvm::ArrayRef<llvm::Constant *> indicesRef(indices);
+    llvm::Constant *mask = llvm::ConstantVector::get(indicesRef);
```

Comment:
Maybe it would be more efficient to create mask as
```
llvm::Constant *mask = llvm::ConstantDataVector::get(B.getContext(), indices);
```
where indices is `std::vector<unsigned> indices(N * 2);`

Reply:
Fixed